### PR TITLE
Add a WFP provider object.

### PR DESCRIPTION
**Pull request by shankarseal**
## Description

Windows HLK tests for WFP requires drivers to add a provider object to identify the vendor. The changes were to add the provider to pass the HLK tests.

## Testing

CI.

## Documentation

Not required.

### CODE REVIEW

File: netebpfext/user/fwp_um.h [hunk 1]
```diff
@@ -171,6 +171,12 @@ typedef class _fwp_engine
         return return_value;
     }
 
+    _Requires_lock_not_held_(this->lock) void add_fwpm_provider(_In_ const FWPM_PROVIDER* sub_layer)
```
Comment (saxena-anurag) [inline netebpfext/user/fwp_um.h:174]:
```suggestion
    _Requires_lock_not_held_(this->lock) void add_fwpm_provider(_In_ const FWPM_PROVIDER* provider)
```

File: netebpfext/user/fwp_um.h [hunk 2]
```diff
@@ -171,6 +171,12 @@ typedef class _fwp_engine
...
+    {
+        UNREFERENCED_PARAMETER(sub_layer);
```
Comment (saxena-anurag) [inline netebpfext/user/fwp_um.h:176]:
```suggestion
        UNREFERENCED_PARAMETER(provider);
```

File: netebpfext/net_ebpf_ext.c
Comment (gtrevi) [inline netebpfext/net_ebpf_ext.c:1]: Add global API string constants:

```
static const char *FWPM_PROVIDER_ADD = "FwpmProviderAdd";
static const char *FWPM_SUBLAYER_ADD = "FwpmSubLayerAdd";
static const char *FWPM_ENGINE_OPEN = "FwpmEngineOpen";
static const char *FWPM_TRANSACTION_BEGIN = "FwpmTransactionBegin";
static const char *FWPM_TRANSACTION_COMMIT = "FwpmTransactionCommit";
static const char *FWPS_INJECTION_HANDLE_CREATE = "FwpsInjectionHandleCreate";
```

File: netebpfext/net_ebpf_ext.c [hunk 3]
```diff
@@ -606,34 +609,33 @@ net_ebpf_extension_initialize_wfp_components(_Inout_ void* device_object)
     session.flags = FWPM_SESSION_FLAG_DYNAMIC;
 
     status = FwpmEngineOpen(NULL, RPC_C_AUTHN_WINNT, NULL, &session, &_fwp_engine_handle);
-    if (!NT_SUCCESS(status)) {
-        NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(NET_EBPF_EXT_TRACELOG_KEYWORD_ERROR, "FwpmEngineOpen", status);
-        goto Exit;
-    }
+    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS("FwpmEngineOpen", status);
```
Comment (gtrevi) [inline netebpfext/net_ebpf_ext.c:612]:
```suggestion
    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS(FWPM_ENGINE_OPEN, status);
```

File: netebpfext/net_ebpf_ext.c [hunk 4]
```diff
@@ -606,34 +609,33 @@ net_ebpf_extension_initialize_wfp_components(_Inout_ void* device_object)
...
     is_engined_opened = TRUE;
 
     status = FwpmTransactionBegin(_fwp_engine_handle, 0);
-    if (!NT_SUCCESS(status)) {
-        NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(NET_EBPF_EXT_TRACELOG_KEYWORD_ERROR, "FwpmTransactionBegin", status);
-        goto Exit;
-    }
+    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS("FwpmTransactionBegin", status);
```
Comment (gtrevi) [inline netebpfext/net_ebpf_ext.c:616]:
```suggestion
    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS(FWPM_TRANSACTION_BEGIN, status);
```

File: netebpfext/net_ebpf_ext.c [hunk 5]
```diff
@@ -606,34 +609,33 @@ net_ebpf_extension_initialize_wfp_components(_Inout_ void* device_object)
...
     is_in_transaction = TRUE;
 
+    // Create the WFP provider.
+    ebpf_wfp_provider.displayData.name = L"Microsoft Corporation";
+    ebpf_wfp_provider.displayData.description = L"Windows Networking eBPF Extension";
+    ebpf_wfp_provider.providerKey = EBPF_WFP_PROVIDER;
+    status = FwpmProviderAdd(_fwp_engine_handle, &ebpf_wfp_provider, NULL);
+    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS("FwpmProviderAdd", status);
```
Comment (gtrevi) [inline netebpfext/net_ebpf_ext.c:624]:
```suggestion
    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS(FWPM_PROVIDER_ADD, status);
```

File: netebpfext/net_ebpf_ext.c [hunk 6]
```diff
@@ -606,34 +609,33 @@ net_ebpf_extension_initialize_wfp_components(_Inout_ void* device_object)
...
+        ebpf_hook_sub_layer.providerKey = (GUID*)&EBPF_WFP_PROVIDER;
         ebpf_hook_sub_layer.flags = _net_ebpf_ext_sublayers[index].flags;
         ebpf_hook_sub_layer.weight = _net_ebpf_ext_sublayers[index].weight;
 
         status = FwpmSubLayerAdd(_fwp_engine_handle, &ebpf_hook_sub_layer, NULL);
-        if (!NT_SUCCESS(status)) {
-            NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(NET_EBPF_EXT_TRACELOG_KEYWORD_ERROR, "FwpmSubLayerAdd", status);
-            goto Exit;
-        }
+        NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS("FwpmSubLayerAdd", status);
```
Comment (gtrevi) [inline netebpfext/net_ebpf_ext.c:638]:
```suggestion
        NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS(FWPM_SUBLAYER_ADD, status);
```

File: netebpfext/net_ebpf_ext.c [hunk 7]
```diff
@@ -649,18 +651,12 @@ net_ebpf_extension_initialize_wfp_components(_Inout_ void* device_object)
     }
 
     status = FwpmTransactionCommit(_fwp_engine_handle);
-    if (!NT_SUCCESS(status)) {
-        NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(NET_EBPF_EXT_TRACELOG_KEYWORD_ERROR, "FwpmTransactionCommit", status);
-        goto Exit;
-    }
+    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS("FwpmTransactionCommit", status);
```
Comment (gtrevi) [inline netebpfext/net_ebpf_ext.c:654]:
```suggestion
    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS(FWPM_TRANSACTION_COMMIT, status);
```

File: netebpfext/net_ebpf_ext.c [hunk 8]
```diff
@@ -649,18 +651,12 @@ net_ebpf_extension_initialize_wfp_components(_Inout_ void* device_object)
...
     is_in_transaction = FALSE;
 
     // Create L2 injection handle.
     status = FwpsInjectionHandleCreate(AF_LINK, FWPS_INJECTION_TYPE_L2, &_net_ebpf_ext_l2_injection_handle);
-    if (!NT_SUCCESS(status)) {
-        NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(NET_EBPF_EXT_TRACELOG_KEYWORD_ERROR, "FwpsInjectionHandleCreate", status);
-        goto Exit;
-    }
+    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS("FwpsInjectionHandleCreate", status);
```
Comment (gtrevi) [inline netebpfext/net_ebpf_ext.c:659]:
```suggestion
    NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS(FWPS_INJECTION_HANDLE_CREATE, status);
```

**Comment by dthaler** [comment]
> ## Testing
> No.
>
> ## Documentation
> No.

https://github.com/microsoft/ebpf-for-windows/blob/main/CONTRIBUTING.md#contributing-code says:
> For all new Pull Requests the following rules apply:
>
> * Existing tests should continue to pass.
> * [Tests](https://github.com/microsoft/ebpf-for-windows/blob/main/docs/AutomatedTests.md) need to be provided for every bug/feature that is completed.
> * Documentation needs to be provided for every feature that is end-user visible.

so saying "Testing: No" is not allowed.   You need to at minimum say it's covered by existing tests (this isn't), or by manual testing or whatever.  The description mentions HLK tests so I would expect the Testing section to talk about that.

**Review by dthaler — COMMENTED** [review]
See comment above.

**Comment by gtrevi** [comment]
> > ## Testing
> > No.
> > ## Documentation
> > No.
> 
> https://github.com/microsoft/ebpf-for-windows/blob/main/CONTRIBUTING.md#contributing-code says:
> 
> > For all new Pull Requests the following rules apply:
> > 
> > * Existing tests should continue to pass.
> > * [Tests](https://github.com/microsoft/ebpf-for-windows/blob/main/docs/AutomatedTests.md) need to be provided for every bug/feature that is completed.
> > * Documentation needs to be provided for every feature that is end-user visible.
> 
> so saying "Testing: No" is not allowed. You need to at minimum say it's covered by existing tests (this isn't), or by manual testing or whatever. The description mentions HLK tests so I would expect the Testing section to talk about that. But we really want tests to be part of CI/CD, so either you should argue that HLK tests can be added to the github workflow, or else you should add a unit test that covers this functionality.

This mod will be covered by manual HLK tests (which required this addition), so we could just add that.

File: netebpfext/net_ebpf_ext.c
Comment (shankarseal) [reply netebpfext/net_ebpf_ext.c:1]: IMHO that would be an overkill as these are needed only for tracing API names, which are not going to change. So literal constant should be OK.

File: netebpfext/net_ebpf_ext.c (hunk 3 above)
Comment (shankarseal) [reply netebpfext/net_ebpf_ext.c:612]: Resolving this and other similar comments based on the first comment on this.

**Review by gtrevi — APPROVED** [review]

**Comment by shankarseal** [comment]
> > ## Testing
> > No.
> > ## Documentation
> > No.
> 
> https://github.com/microsoft/ebpf-for-windows/blob/main/CONTRIBUTING.md#contributing-code says:
> 
> > For all new Pull Requests the following rules apply:
> > 
> > * Existing tests should continue to pass.
> > * [Tests](https://github.com/microsoft/ebpf-for-windows/blob/main/docs/AutomatedTests.md) need to be provided for every bug/feature that is completed.
> > * Documentation needs to be provided for every feature that is end-user visible.
> 
> so saying "Testing: No" is not allowed. You need to at minimum say it's covered by existing tests (this isn't), or by manual testing or whatever. The description mentions HLK tests so I would expect the Testing section to talk about that. But we really want tests to be part of CI/CD, so either you should argue that HLK tests can be added to the github workflow, or else you should add a unit test that covers this functionality.

I wanted to write "N/A" - not sure why I wrote the rule "No". Sorry about that.

**Comment by dthaler** [comment]
> I wanted to write "N/A" - not sure why I wrote the rule "No". Sorry about that.

Testing should never say "N/A".  It's always applicable.  For things like doc-only changes "No impact." would be the answer, but this is not a doc-only change.

**Comment by shankarseal** [comment]
> > I wanted to write "N/A" - not sure why I wrote the rule "No". Sorry about that.
> 
> Testing should never say "N/A". It's always applicable. For things like doc-only changes "No impact." would be the answer, but this is not a doc-only change.

Done. please approve.

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**